stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,236 @@
+#ifndef STAN_MATH_REV_SCAL_FUNCTOR_REDUCE_SUM_HPP
+#define STAN_MATH_REV_SCAL_FUNCTOR_REDUCE_SUM_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/rev/core/deep_copy_vars.hpp>
+#include <stan/math/rev/core/accumulate_adjoints.hpp>
+#include <stan/math/rev/core/count_vars.hpp>
+#include <stan/math/rev/core/save_varis.hpp>
+#include <stan/math/rev/core.hpp>
+#include <tbb/task_arena.h>
+#include <tbb/parallel_reduce.h>
+#include <tbb/blocked_range.h>
+
+#include <tuple>
+#include <vector>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Var specialization of reduce_sum_impl
+ *
+ * @tparam ReduceFunction Type of reducer function
+ * @tparam ReturnType Must be var
+ * @tparam Vec Type of sliced argument
+ * @tparam Args Types of shared arguments
+ */
+template <typename ReduceFunction, typename ReturnType, typename Vec,
+          typename... Args>
+struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
+                       Vec, Args...> {
+  /**
+   * Internal object meeting the Imperative form requirements of
+   * `tbb::parallel_reduce`
+   *
+   * @note see link [here](https://tinyurl.com/vp7xw2t) for requirements.
+   */
+  struct recursive_reducer {
+    size_t per_job_sliced_terms_;
+    size_t num_shared_terms_;  // Number of terms shared across threads
+    double* sliced_partials_;  // Points to adjoints of the partial calculations
+    Vec vmapped_;
+    std::ostream* msgs_;
+    std::tuple<Args...> args_tuple_;
+    double sum_{0.0};
+    Eigen::VectorXd args_adjoints_{0};
+
+    template <typename VecT, typename... ArgsT>
+    recursive_reducer(size_t per_job_sliced_terms, size_t num_shared_terms,
+                      double* sliced_partials, VecT&& vmapped,
+                      std::ostream* msgs, ArgsT&&... args)
+        : per_job_sliced_terms_(per_job_sliced_terms),
+          num_shared_terms_(num_shared_terms),
+          sliced_partials_(sliced_partials),
+          vmapped_(std::forward<VecT>(vmapped)),
+          msgs_(msgs),
+          args_tuple_(std::forward<ArgsT>(args)...) {}
+
+    /*
+     * This is the copy operator as required for tbb::parallel_reduce
+     *   Imperative form. This requires the reduced values (sum_ and
+     * arg_adjoints_) be reset to zero.
+     */
+    recursive_reducer(recursive_reducer& other, tbb::split)
+        : per_job_sliced_terms_(other.per_job_sliced_terms_),
+          num_shared_terms_(other.num_shared_terms_),
+          sliced_partials_(other.sliced_partials_),
+          vmapped_(other.vmapped_),
+          msgs_(other.msgs_),
+          args_tuple_(other.args_tuple_) {}
```

Comment:
Is `tbb::split` supposed to do something here?

Reply:
That's used by the TBB API to discriminate the constructor from the copy constructor if I recall right. So it's needed by the TBB.